python/ray/_private/runtime_env/validation.py

```diff
@@ -116,9 +116,33 @@ def parse_and_validate_conda(conda: Union[str, dict]) -> Union[str, dict]:
 
 
 def _rewrite_pip_list_ray_libraries(pip_list: List[str]) -> List[str]:
+    """Remove Ray and replace Ray libraries with their dependencies.
+
+    The `pip` field of runtime_env installs packages into the current
+    environment, inheriting the existing environment.  If users want to
+    use Ray libraries like `ray[serve]` in their job, they must include
+    `ray[serve]` in their `runtime_env` `pip` field.  However, without this
+    function, the Ray installed at runtime would take precedence over the
+    Ray that exists in the cluster, which would lead to version mismatch
+    issues.
+
+    To work around this, this function deletes Ray from the input `pip_list`
+    if it's specified without any libraries (e.g. "ray" or "ray>1.4"). If
+    a Ray library is specified (e.g. "ray[serve]"), it is replaced by
+    its dependencies (e.g. "uvicorn", ...).
+
+    """
     result = []
     for specifier in pip_list:
-        requirement = Requirement.parse(specifier)
+        try:
+            requirement = Requirement.parse(specifier)
+        except Exception:
+            # Some lines in a pip_list might not be requirements but
+            # rather options for `pip`; e.g. `--extra-index-url MY_INDEX`.
+            # Requirement.parse would raise an InvalidRequirement in this
+            # case.  Since we are only interested in lines specifying Ray
+            # or its libraries, we should just skip this line.
+            continue
```

Comment:
just to be clear, these options will still be passed through to the final call to create the env, right?

Reply:
Oh wow, good catch.  Let me fix this and fix the test so that it catches this.

Reply:
Fixed